### PR TITLE
Opensearch Source Boilerplate code for #1985,2264.

### LINKED ISSUE #1985: OpenSearch source plugin
**Is your feature request related to a problem? Please describe.**

Some users have looked for a way to migrate data from one OpenSearch cluster to another. One of the missing components here would be the ability to retrieve data from OpenSearch. Other users have expressed a need to transform data from OpenSearch clusters.

**Describe the solution you'd like**

Include an OpenSearch source plugin for Data Prepper. It will need some of the following configurations.

* Connection configurations similar to the `opensearch` sink.
* Source index or indices. This should also support dynamic indices based on date.
* Possibly query options to filter down the data. If not specified, all data in the index would be used.
* Schedule configurations for reading data.

**Additional context**

This should be similar to the `logstash-input-opensearch-plugin` provided in the OpenSearch project.

https://opensearch.org/blog/community/2022/05/introducing-logstash-input-opensearch-plugin-for-opensearch/
https://github.com/opensearch-project/logstash-input-opensearch

### PR DESCRIPTION
**Pull request by mallikagogoi7**
### Description
Opensearch Source Boilerplate code
 
### Issues Resolved
#1985,2264
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/PITBuilder.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.source.opensearch.model;
+
+import org.opensearch.client.util.ObjectBuilder;
+
+import static java.util.Objects.requireNonNull;
+
+public  class PITBuilder implements ObjectBuilder<PITRequest> {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/PITBuilder.java:12]: Please rename this class to `PitBuilder`.

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/PITRequest.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import org.opensearch.client.json.JsonpSerializable;
+import org.opensearch.client.opensearch._types.ErrorResponse;
+import org.opensearch.client.transport.Endpoint;
+import org.opensearch.client.transport.endpoints.SimpleEndpoint;
+import org.opensearch.dataprepper.plugins.source.opensearch.codec.JacksonValueParser;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class PITRequest implements JsonpSerializable {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/PITRequest.java:20]: Please rename to `PitRequest`.

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/PITResponse.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.source.opensearch.model;
+
+import org.opensearch.client.json.JsonpDeserializer;
+
+public class PITResponse  {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/PITResponse.java:10]: Please update to `PitResponse`.

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/OpenSearchClientBuilder.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+/**
+ * used for creating connection
+ */
+public class OpenSearchClientBuilder {
+
+    public OpenSearchClient createOpenSearchClient(final URL url){
+        return null;
+    }
+
+    public ElasticsearchClient createElasticSearchCLient(final URL url) {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/OpenSearchClientBuilder.java:20]: Correct the case of the `L` in `CLient` to `Client`.

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/OpenSearchSourceService.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                                   final BufferAccumulator bufferAccumulator){
+        this.sourceConfig = sourceConfig;
+        this.hostsService = hostsService;
+        this.openSearchService = openSearchService;
+        this.elasticSearchService =elasticSearchService;
+        this.bufferAccumulator = bufferAccumulator;
+    }
+
+    public void processHosts(){
+        sourceConfig.getHosts().forEach(host ->{
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/OpenSearchSourceService.java:45]: The hosts should be shared and provided to the OpenSearch Java client.

These `hosts` represent the different hosts which make up the OpenSearch cluster. They should not have their own schedules.

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/DeletePITResponse.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+package org.opensearch.dataprepper.plugins.source.opensearch.model;
+
+import java.io.Serializable;
+import java.util.List;
+
+/**
+ * Reference DeletePITResponse Object class
+ */
+public class DeletePITResponse implements Serializable {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/DeletePITResponse.java:14]: Please rename to `DeletePitResponse`.

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/DeletePITResponse.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+        public Boolean getSuccessful() {
+            return successful;
+        }
+
+        public void setSuccessful(Boolean successful) {
+            this.successful = successful;
+        }
+
+        public String getPit_id() {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/DeletePITResponse.java:32]: Should be `getPitId`.

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/DeletePITResponse.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            return pit_id;
+        }
+
+        public void setPit_id(String pit_id) {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/DeletePITResponse.java:36]: Should be `setPitId`

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/Hit.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonPropertyOrder({
+    "_index",
+    "_id",
+    "_score",
+    "_source",
+    "sort"
+})
+public class Hit {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/Hit.java:28]: Is this different from `org.opensearch.client.opensearch.core.search.Hit`? Can we not use that?

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/Hits.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.List;
+import java.util.Map;
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonPropertyOrder({
+    "total",
+    "max_score",
+    "hits"
+})
+public class Hits {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/Hits.java:25]: The `org.opensearch.client.opensearch.core.search` package has some models for hits. Can we use any of those?

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/PITRequest.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        return queryParameters;
+    }
+
+    private Map<String,String> params = new HashMap<>();
+
+    public void setParams(Map<String, String> params) {
+        this.params = params;
+    }
+
+    public static final Endpoint<PITRequest, Map, ErrorResponse> ENDPOINT =
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/PITRequest.java:53]: This should have unit tests.

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/PITRequest.java (hunk 2 above)
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/PITRequest.java:20]: This whole class can have unit tests.

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/PITResponse.java [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+    public static JsonpDeserializer<PITResponse> PARSER;
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/model/PITResponse.java:12]: Do we need this? And if so, how can we better hide it?

Using public variable fields is often not recommended.

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/service/BackoffService.java [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Random;
+
+/**
+ *  Service that handle retry mechanism in case of any api failure
+ */
+public class BackoffService {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/service/BackoffService.java:16]: Armeria has a class `Backoff`:

```
com.linecorp.armeria.client.retry.Backoff
```

Let's use this.

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/service/OpenSearchService.java [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+    private static final Integer BATCH_SIZE_VALUE = 1000;
+
+    private static final int OPEN_SEARCH_VERSION = 130;
+
+    public List<IndicesRecord> getCatOpenSearchIndices(final OpenSearchClient osClient){
+
+      return null;
+    }
+
+    public String getPITId(final String index){
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/service/OpenSearchService.java:27]: Rename to `getPitId`

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/service/OpenSearchService.java [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+        // if response is not 200 then will call BackoffService for retry
+        return null;
+    }
+
+    public Map<String, Object> searchIndexesByPITId(final String pitId) {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/service/OpenSearchService.java:32]: Rename to `searchIndexesByPitId`

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/service/OpenSearchService.java [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+        // if response is not 200 then will call BackoffService for retry
+        return null;
+    }
+
+    public Map<String, Object> searchIndexesByPITIdForPagination(final String pitId) {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/service/OpenSearchService.java:37]: `searchIndexesByPitIdForPagination`

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/service/OpenSearchService.java [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+
+    public Map<String, Object> scrollIndexesByIndexAndUrl(final String index, final String url) {
+        // if response is not 200 then will call BackoffService for retry
+        return null;
+    }
+
+
+    public boolean deletePITId(String pitId) {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/service/OpenSearchService.java:49]: Rename to `deletePitId`

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/worker/OpenSearchTimerWorker.java [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+        this.sourceConfig = sourceConfig;
+        this.bufferAccumulator = bufferAccumulator;
+        this.serviceInfo = serviceInfo;
+        this.host = host;
+    }
+
+    @Override
+    public void run() {
+        for(int jobCount=1 ; sourceConfig.getSchedulingParameterConfiguration().getJobCount() >= jobCount; jobCount++) {
+            if (OPEN_SEARCH_DISTRIBUTION.equals(serviceInfo.getDistribution())) {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/worker/OpenSearchTimerWorker.java:49]: Let's avoid this condition in this class`.

Please inject a strategy into this class via the constructor. Then you will have a single mechanism for processing indexes.

You will inject a `CommonSearchService` in this class.

This code then becomes:

```
@Override
public void run() {
  for(int jobCount=1 ; sourceConfig.getSchedulingParameterConfiguration().getJobCount() >= jobCount; jobCount++) {
    commonSearchService.processIndexes(serviceInfo.getVersion(),
                        getIndexList(),
                        host,
                        sourceConfig.getSearchConfiguration().getBatchSize());
```

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/service/OpenSearchService.java [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+import org.opensearch.client.opensearch.OpenSearchClient;
+import org.opensearch.client.opensearch.cat.indices.IndicesRecord;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * OpenSearch service related implementation
+ */
+public class OpenSearchService {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/service/OpenSearchService.java:16]: Please create an interface to abstract this and `ElasticSearchService`.

```
public interface CommonSearchService {
  void processIndexes(final Integer version,
                               final String indexList,
                               final String url,
                               final Integer batchSize)
}
```

I'm open to a better name than `CommonSearchService`.

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/service/OpenSearchService.java [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+
+        return false;
+    }
+
+    public boolean deleteScrollId(String scrollId) {
+
+        return false;
+    }
+
+    public void processIndexes(final Integer version,
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/service/OpenSearchService.java:59]: This needs unit testing.

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/worker/OpenSearchTimerWorker.java [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+            } else {
+                elasticSearchService.processIndexes(serviceInfo.getVersion(),
+                        getIndexList(),
+                        host,
+                        sourceConfig.getSearchConfiguration().getBatchSize());
+            }
+        }
+    }
+
+    private String getIndexList()
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/worker/OpenSearchTimerWorker.java:63]: Let's move this into a different class specifically for determining the indexes.

You can also add quality unit tests to such a class.

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/worker/OpenSearchTimerWorker.java [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+import org.opensearch.dataprepper.plugins.source.opensearch.OpenSearchSourceConfiguration;
+import org.opensearch.dataprepper.plugins.source.opensearch.model.ServiceInfo;
+import org.opensearch.dataprepper.plugins.source.opensearch.service.ElasticSearchService;
+import org.opensearch.dataprepper.plugins.source.opensearch.service.OpenSearchService;
+
+import java.util.List;
+import java.util.TimerTask;
+import java.util.stream.Collectors;
+
+public class OpenSearchTimerWorker extends TimerTask {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/worker/OpenSearchTimerWorker.java:17]: This class needs unit tests.

File: data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/codec/JacksonValueParser.java [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import jakarta.json.stream.JsonParser;
+import org.opensearch.client.json.JsonpDeserializerBase;
+import org.opensearch.client.json.JsonpMapper;
+import org.opensearch.client.json.jackson.JacksonJsonpParser;
+import java.io.IOException;
+import java.util.EnumSet;
+
+public class JacksonValueParser<T> extends JsonpDeserializerBase<T> {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/codec/JacksonValueParser.java:16]: This class needs unit tests.

**Review by dlvenable — CHANGES_REQUESTED** [review]
Thanks for this contribution @mallikagogoi7 !

**Comment by cmanning09** [comment]
@mallikagogoi7, thanks for your contribution. We have restructured this plugin and this change is no longer needed. I am going to close this contribution.

**Pull request closed without merging**